lib/wallet/coinselector.js

```diff
@@ -0,0 +1,896 @@
+      }
+    }
+
+    if (options.getAccount) {
+      assert(typeof options.getAccount === 'function');
+      this.getAccount = options.getAccount;
+    }
+
+    if (options.inputs) {
+      assert(Array.isArray(options.inputs));
+      for (let i = 0; i < options.inputs.length; i++) {
+        const prevout = options.inputs[i];
+        assert(prevout && typeof prevout === 'object');
+        const {hash, index} = prevout;
+        assert(Buffer.isBuffer(hash));
+        assert(typeof index === 'number');
+        this.inputs.set(Outpoint.toKey(hash, index), i);
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * Attempt to inject existing inputs.
+   * @private
+   */
+
+  injectInputs() {
+    if (this.tx.inputs.length > 0) {
+      for (let i = 0; i < this.tx.inputs.length; i++) {
+        const {prevout} = this.tx.inputs[i];
+        this.inputs.set(prevout.toKey(), i);
+      }
+    }
+  }
+
+  /**
+   * Initialize the selector with coins to select from.
+   * @param {Coin[]} coins
+   */
+
+  async init(coins) {
+    this.coins = coins.slice();
+    this.outputValue = this.tx.getOutputValue();
+    this.index = 0;
+    this.chosen = [];
+    this.change = 0;
+    this.fee = CoinSelector.MIN_FEE;
+    this.tx.inputs.length = 0;
+
+    switch (this.selection) {
+      case 'all':
+      case 'random':
+        this.coins.sort(sortRandom);
+        break;
+      case 'age':
+        this.coins.sort(sortAge);
+        break;
+      case 'value':
+        this.coins.sort(sortValue);
+        break;
+      default:
+        throw new FundingError(`Bad selection type: ${this.selection}.`);
+    }
+
+    if (!this.useSelectEstimate) {
+      for (let i = 0, n = coins.length; i < n; i++) {
+        const coin = coins[i];
+        if (this.isSpendable(coin)) {
+          const utxo = new UTXO();
+          utxo.index = i;
+          utxo.spendingSize = await coin.estimateSpendingSize(this.getAccount);
+          utxo.effectiveValue = coin.value - this.getFee(utxo.spendingSize);
+          this.utxos.push(utxo);
+        }
+      }
+      this.utxos.sort((a, b) => b.effectiveValue - a.effectiveValue);
+    }
+  }
+
+  /**
+   * Calculate total value required.
+   * @returns {Amount}
+   */
+
+  total() {
+    if (this.subtractFee)
+      return this.outputValue;
+    return this.outputValue + this.fee;
+  }
+
+  /**
+   * Test whether the selector has
+   * completely funded the transaction.
+   * @returns {Boolean}
+   */
+
+  isFull() {
+    return this.tx.getInputValue() >= this.total();
+  }
+
+  /**
+   * Test whether a coin is spendable
+   * with regards to the options.
+   * @param {Coin} coin
+   * @returns {Boolean}
+   */
+
+  isSpendable(coin) {
+    if (this.tx.view.hasEntry(coin))
+      return false;
+
+    if (this.height === -1)
+      return true;
+
+    if (coin.coinbase) {
+      if (coin.height === -1)
+        return false;
+
+      if (this.height + 1 < coin.height + consensus.COINBASE_MATURITY)
```

Comment:
the if statement can be simplified to ` return this.height + 1 >= coin.height + consensus.COINBASE_MATURITY;`